File: backend/src/models/Order.js
const { v4: uuidv4 } = require('uuid');
const { parseRowData, decodeEscapedUTF8 } = require('../utils/helpers');

class Order {
  constructor(data) {
    this.id = data.id || `order#${uuidv4()}`;
    this.userId = data.userId || '';
    this.storeId = data.storeId || '';
    this.items = data.items || [];
    this.subtotal = data.subtotal || 0;
    this.tax = data.tax || 0;
    this.deliveryFee = data.deliveryFee || 0;
    this.discount = data.discount || 0;
    this.total = data.total || 0;
    this.status = data.status || 'pending'; // pending, confirmed, preparing, delivering, completed, cancelled
    this.paymentMethod = data.paymentMethod || 'COD'; // COD, card, momo, zalopay
    this.paymentStatus = data.paymentStatus || 'pending'; // pending, paid, failed, refunded
    this.deliveryAddress = data.deliveryAddress || {};
    this.notes = data.notes || '';
    this.orderTime = data.orderTime || new Date().toISOString();
    this.confirmedTime = data.confirmedTime || null;
    this.completedTime = data.completedTime || null;
    this.cancelledTime = data.cancelledTime || null;
    this.cancelReason = data.cancelReason || null;
    this.promotionCode = data.promotionCode || null;
  }

  canBeCancelled() {
    return ['pending', 'confirmed'].includes(this.status);
  }

  toJSON() {
    return {
      id: this.id || '',
      userId: this.userId || '',
      storeId: this.storeId || '',
      items: this.items || [],
      subtotal: this.subtotal || 0,
      tax: this.tax || 0,
      deliveryFee: this.deliveryFee || 0,
      discount: this.discount || 0,
      total: this.total || 0,
      status: this.status || 'pending',
      paymentMethod: this.paymentMethod || 'COD',
      paymentStatus: this.paymentStatus || 'pending',
      deliveryAddress: this.deliveryAddress || {},
      notes: this.notes || '',
      orderTime: this.orderTime || new Date().toISOString(),
      confirmedTime: this.confirmedTime || null,
      completedTime: this.completedTime || null,
      cancelledTime: this.cancelledTime || null,
      cancelReason: this.cancelReason || null,
      promotionCode: this.promotionCode || null,
    };
  }

  static fromBigtableRow(row, rowData) {
    // Use parseRowData helper to ensure proper UTF-8 decoding
    const parsedData = parseRowData(rowData);
    
    const items = [];
    
    // Parse items from Bigtable
    Object.keys(parsedData).forEach(key => {
      if (key.startsWith('item_')) {
        try {
          // Properly decode UTF-8 for item data
          let itemValue = parsedData[key];
          itemValue = decodeEscapedUTF8(itemValue);
          items.push(JSON.parse(itemValue || '{}'));
        } catch (error) {
          console.error(`Error parsing item ${key}:`, error);
        }
      }
    });

    return new Order({
      id: row.id || '',
      userId: decodeEscapedUTF8(parsedData.userId) || '',
      storeId: decodeEscapedUTF8(parsedData.storeId) || '',
      items: items,
      subtotal: parseFloat(parsedData.subtotal) || 0,
      tax: parseFloat(parsedData.tax) || 0,
      deliveryFee: parseFloat(parsedData.deliveryFee) || 0,
      discount: parseFloat(parsedData.discount) || 0,
      total: parseFloat(parsedData.total) || 0,
      status: parsedData.status || 'pending',
      paymentMethod: parsedData.paymentMethod || 'COD',
      paymentStatus: parsedData.paymentStatus || 'pending',
      deliveryAddress: parsedData.deliveryAddress ? JSON.parse(parsedData.deliveryAddress || '{}') : {},
      notes: decodeEscapedUTF8(parsedData.notes) || '',
      orderTime: parsedData.orderTime || new Date().toISOString(),
      confirmedTime: parsedData.confirmedTime || null,
      completedTime: parsedData.completedTime || null,
      cancelledTime: parsedData.cancelledTime || null,
      cancelReason: parsedData.cancelReason || null,
      promotionCode: parsedData.promotionCode || null,
    });
  }

  // Generate row key with reversed timestamp for efficient recent-first queries
  static generateRowKey(orderId) {
    const timestamp = Date.now();
    const reversedTimestamp = 9999999999999 - timestamp;
    return `order#${reversedTimestamp}#${orderId}`;
  }
}

module.exports = Order;

File: lib/models/order.dart
import 'cart_item.dart';
import 'store.dart';

class Order {
  final String id;
  final String userId;
  final Store store;
  final List<CartItem> items;
  final double subtotal;
  final double tax;
  final double deliveryFee;
  final double discount;
  final double total;
  final OrderStatus status;
  final PaymentMethod paymentMethod;
  final PaymentStatus paymentStatus;
  final DeliveryMethod deliveryMethod;
  final Map<String, dynamic>? deliveryAddress;
  final DateTime orderTime;
  final DateTime? pickupTime;
  final DateTime? completedTime;
  final String? notes;
  final String? promotionCode;

  Order({
    required this.id,
    required this.userId,
    required this.store,
    required this.items,
    required this.subtotal,
    required this.tax,
    this.deliveryFee = 0.0,
    this.discount = 0.0,
    required this.total,
    required this.status,
    required this.paymentMethod,
    required this.paymentStatus,
    this.deliveryMethod = DeliveryMethod.pickup,
    this.deliveryAddress,
    required this.orderTime,
    this.pickupTime,
    this.completedTime,
    this.notes,
    this.promotionCode,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    // Parse DateTime - handle null and empty strings
    DateTime? parseDateTime(dynamic value) {
      if (value == null) return null;
      if (value is String && value.isEmpty) return null;
      try {
        return DateTime.parse(value.toString());
      } catch (e) {
        return null;
      }
    }
    
    // Parse deliveryAddress - handle both object and null
    Map<String, dynamic>? parseDeliveryAddress(dynamic value) {
      if (value == null) return null;
      if (value is Map<String, dynamic>) return value;
      if (value is Map) return Map<String, dynamic>.from(value);
      return null;
    }
    
    return Order(
      id: json['id']?.toString() ?? '',
      userId: json['userId']?.toString() ?? '',
      store: Store.fromJson(json['store'] as Map<String, dynamic>),
      items: (json['items'] as List<dynamic>)
          .map((e) => CartItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      subtotal: (json['subtotal'] as num?)?.toDouble() ?? 0.0,
      tax: (json['tax'] as num?)?.toDouble() ?? 0.0,
      deliveryFee: (json['deliveryFee'] as num?)?.toDouble() ?? 0.0,
      discount: (json['discount'] as num?)?.toDouble() ?? 0.0,
      total: (json['total'] as num?)?.toDouble() ?? 0.0,
      status: OrderStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => OrderStatus.pending,
      ),
      paymentMethod: PaymentMethod.values.firstWhere(
        (e) => e.name == json['paymentMethod'],
        orElse: () => PaymentMethod.card,
      ),
      paymentStatus: PaymentStatus.values.firstWhere(
        (e) => e.name == json['paymentStatus'],
        orElse: () => PaymentStatus.pending,
      ),
      deliveryMethod: DeliveryMethod.values.firstWhere(
        (e) => e.name == json['deliveryMethod'],
        orElse: () => DeliveryMethod.pickup,
      ),
      deliveryAddress: parseDeliveryAddress(json['deliveryAddress']),
      orderTime: parseDateTime(json['orderTime']) ?? DateTime.now(),
      pickupTime: parseDateTime(json['pickupTime']),
      completedTime: parseDateTime(json['completedTime']),
      notes: json['notes']?.toString(),
      promotionCode: json['promotionCode']?.toString(),
    );
  }

  Map<String, dynamic> toJson() {
    // Validate all items have valid productId
    for (final item in items) {
      if (item.product.id.isEmpty) {
        throw Exception('Product ID cannot be empty for item: ${item.product.name}');
      }
    }
    
    return {
      'id': id,
      'userId': userId,
      'storeId': store.id, // Send storeId instead of full store object
      'items': items.map((e) => {
        'productId': e.product.id, // Ensure product.id is not null
        'quantity': e.quantity,
        'size': e.size,
        'options': e.selectedOptions,
      }).toList(),
      'subtotal': subtotal,
      'tax': tax,
      'total': total,
      'status': status.name,
      'paymentMethod': paymentMethod.name,
      'paymentStatus': paymentStatus.name,
      'deliveryMethod': deliveryMethod.name,
      'orderTime': orderTime.toIso8601String(),
      'pickupTime': pickupTime?.toIso8601String(),
      'completedTime': completedTime?.toIso8601String(),
      'notes': notes,
    };
  }

  Order copyWith({
    String? id,
    String? userId,
    Store? store,
    List<CartItem>? items,
    double? subtotal,
    double? tax,
    double? deliveryFee,
    double? discount,
    double? total,
    OrderStatus? status,
    PaymentMethod? paymentMethod,
    PaymentStatus? paymentStatus,
    DeliveryMethod? deliveryMethod,
    Map<String, dynamic>? deliveryAddress,
    DateTime? orderTime,
    DateTime? pickupTime,
    DateTime? completedTime,
    String? notes,
    String? promotionCode,
  }) {
    return Order(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      store: store ?? this.store,
      items: items ?? this.items,
      subtotal: subtotal ?? this.subtotal,
      tax: tax ?? this.tax,
      deliveryFee: deliveryFee ?? this.deliveryFee,
      discount: discount ?? this.discount,
      total: total ?? this.total,
      status: status ?? this.status,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      paymentStatus: paymentStatus ?? this.paymentStatus,
      deliveryMethod: deliveryMethod ?? this.deliveryMethod,
      deliveryAddress: deliveryAddress ?? this.deliveryAddress,
      orderTime: orderTime ?? this.orderTime,
      pickupTime: pickupTime ?? this.pickupTime,
      completedTime: completedTime ?? this.completedTime,
      notes: notes ?? this.notes,
      promotionCode: promotionCode ?? this.promotionCode,
    );
  }
}

enum OrderStatus {
  pending,
  confirmed,
  preparing,
  ready,
  completed,
  cancelled,
}

enum PaymentMethod {
  card,
  cash,
  momo,
  zalopay,
}

enum PaymentStatus {
  pending,
  paid,
  failed,
  refunded,
}

enum DeliveryMethod {
  pickup,    // Nhận tại cửa hàng
  delivery,  // Giao tận nơi
}

